registration/include/pcl/registration/impl/ndt_omp.hpp

```diff
@@ -0,0 +1,883 @@
+    h_ang.row(11).noalias() = Eigen::Vector4f((cx * cy * sz), (cx * cy * cz), 0, 0.0f); // e3
+
+    h_ang.row(12).noalias() = Eigen::Vector4f((-cy * cz), (cy * sz), 0, 0.0f); // f1
+    h_ang.row(13).noalias() = Eigen::Vector4f((-cx * sz - sx * sy * cz), (-cx * cz + sx * sy * sz), 0, 0.0f); // f2
+    h_ang.row(14).noalias() = Eigen::Vector4f((-sx * sz + cx * sy * cz), (-cx * sy * sz - sx * cz), 0, 0.0f); // f3
+  }
+}
+
+template <typename PointSource, typename PointTarget> void
+NormalDistributionsTransformOMP<PointSource, PointTarget>::computePointDerivatives(Eigen::Vector3d& x, Eigen::Matrix<float, 4, 6>& point_gradient, Eigen::Matrix<float, 24, 6>& point_hessian, bool compute_hessian) const
+{
+  Eigen::Vector4f x4(x[0], x[1], x[2], 0.0f);
+
+  // Calculate first derivative of Transformation Equation 6.17 w.r.t. transform vector p.
+  // Derivative w.r.t. ith element of transform vector corresponds to column i, Equation 6.18 and 6.19 [Magnusson 2009]
+  Eigen::Matrix<float, 8, 1> x_j_ang = j_ang * x4;
+
+  point_gradient(1, 3) = x_j_ang[0];
+  point_gradient(2, 3) = x_j_ang[1];
+  point_gradient(0, 4) = x_j_ang[2];
+  point_gradient(1, 4) = x_j_ang[3];
+  point_gradient(2, 4) = x_j_ang[4];
+  point_gradient(0, 5) = x_j_ang[5];
+  point_gradient(1, 5) = x_j_ang[6];
+  point_gradient(2, 5) = x_j_ang[7];
+
+  if (compute_hessian) {
+    Eigen::Matrix<float, 16, 1> x_h_ang = h_ang * x4;
+
+    // Vectors from Equation 6.21 [Magnusson 2009]
+    Eigen::Vector4f a(0, x_h_ang[0], x_h_ang[1], 0.0f);
+    Eigen::Vector4f b(0, x_h_ang[2], x_h_ang[3], 0.0f);
+    Eigen::Vector4f c(0, x_h_ang[4], x_h_ang[5], 0.0f);
+    Eigen::Vector4f d(x_h_ang[6], x_h_ang[7], x_h_ang[8], 0.0f);
+    Eigen::Vector4f e(x_h_ang[9], x_h_ang[10], x_h_ang[11], 0.0f);
+    Eigen::Vector4f f(x_h_ang[12], x_h_ang[13], x_h_ang[14], 0.0f);
+
+    // Calculate second derivative of Transformation Equation 6.17 w.r.t. transform vector p.
+    // Derivative w.r.t. ith and jth elements of transform vector corresponds to the 3x1 block matrix starting at (3i,j), Equation 6.20 and 6.21 [Magnusson 2009]
+    point_hessian.block<4, 1>((9 / 3) * 4, 3) = a;
+    point_hessian.block<4, 1>((12 / 3) * 4, 3) = b;
+    point_hessian.block<4, 1>((15 / 3) * 4, 3) = c;
+    point_hessian.block<4, 1>((9 / 3) * 4, 4) = b;
+    point_hessian.block<4, 1>((12 / 3) * 4, 4) = d;
+    point_hessian.block<4, 1>((15 / 3) * 4, 4) = e;
+    point_hessian.block<4, 1>((9 / 3) * 4, 5) = c;
+    point_hessian.block<4, 1>((12 / 3) * 4, 5) = e;
+    point_hessian.block<4, 1>((15 / 3) * 4, 5) = f;
+  }
+}
+
+//////////////////////////////////////////////////////////////////////////////////////////////////////////////////////
+template <typename PointSource, typename PointTarget>
+void
+pcl::NormalDistributionsTransformOMP<PointSource, PointTarget>::computePointDerivatives(Eigen::Vector3d& x, Eigen::Matrix<double, 3, 6>& point_gradient, Eigen::Matrix<double, 18, 6>& point_hessian, bool compute_hessian) const
+{
```

Comment:
You need to zero the point_{gradient,hessian} (Though point_jacobian is better name due to symmetry with hessian)